### PR TITLE
BREAKING: Adds rest of the REST and migrate to Typed EventEmitter, better Gateway interfacing

### PR DESCRIPTION
**Pull request by DjDeveloperr**
- Adds complete REST (as left from the https://github.com/harmony-org/harmony/issues/86, some of the REST will be in Helloyunho's PR)
- Migrate to Typed EventEmitter. removeListener and addListener are gone.
- Gateway class (client.gateway) is now initialized in constructor and connected on `client.connect`
- Added Client.reconnect, Client.destroy
- maybe some more things I forgot 😅

### CODE REVIEW

File: src/managers/guilds.ts [hunk 1]
```diff
@@ -48,5 +60,183 @@ export class GuildManager extends BaseManager<GuildPayload, Guild> {
...
+          : undefined,
+      afk_channel_id: options.afkChannelID,
+      afk_timeout: options.afkTimeout,
+      system_channel_id: options.systemChannelID
+    }
+
+    const result: GuildPayload = await this.client.rest.post(GUILDS(), body)
+    const guild = new Guild(this.client, result)
+
+    return guild
```
Comment (AkiaCode) [inline src/managers/guilds.ts]:
```javascript
return new Guild(this.client, result)
```

File: src/managers/guilds.ts [hunk 2]
```diff
@@ -48,5 +60,183 @@ export class GuildManager extends BaseManager<GuildPayload, Guild> {
...
+    }
+
+    const result: GuildPayload = await this.client.rest.patch(
+      GUILD(guild),
+      body
+    )
+
+    if (asRaw) {
+      const guild = new Guild(this.client, result)
+      return guild
```
Comment (AkiaCode) [inline src/managers/guilds.ts]:
```javascript 
return new Guild(this.client, result)
```

**Review by AkiaCode — APPROVED** [review]

File: src/managers/guilds.ts (hunk 1 above)
Comment (DjDeveloperr) [reply src/managers/guilds.ts]: what, did you read how is guild initialized

File: src/models/client.ts [hunk 3]
```diff
@@ -183,12 +189,17 @@ export class Client extends EventEmitter {
...
+    if (options.shard !== undefined) this.shard = options.shard
+    if (options.shardCount !== undefined) this.shardCount = options.shardCount
+
     this.slash = new SlashClient({
       id: () => this.getEstimatedID(),
       client: this,
       enabled: options.enableSlash
     })
 
+    if (options.fetchGatewayInfo === true) this.fetchGatewayInfo = true
```
Comment (Helloyunho) [inline src/models/client.ts:201]: I think it's better to just do like `this.fetchGatewayInfo = options.fetchGatewayInfo ?? false`

File: src/structures/guild.ts [hunk 4]
```diff
@@ -328,18 +331,137 @@ export class Guild extends Base {
...
+
+  /** Sync an integration. */
+  async syncIntegration(id: string): Promise<Guild> {
+    await this.client.rest.api.guilds[this.id].integrations[id].sync.post()
+    return this
+  }
+
+  /** Returns the widget for the guild. */
+  async getWidget(): Promise<GuildWidgetPayload> {
+    return this.client.rest.api.guilds[this.id]['widget.json'].get()
```
Comment (Helloyunho) [inline src/structures/guild.ts:383]: You can ignore this but converting things into camelCase will be good

Comment (DjDeveloperr) [reply src/structures/guild.ts:383]: A separate structure for this would be overkill 🤔

File: src/models/client.ts (hunk 3 above)
Comment (DjDeveloperr) [reply src/models/client.ts:201]: Oh.. yes

File: src/structures/guild.ts (hunk 4 above)
Comment (Helloyunho) [reply src/structures/guild.ts:383]: Yeah but it matches with our linter style and doesn't make code slower

**Review by Helloyunho — APPROVED** [review]
LGTM

**Review by AkiaCode — APPROVED** [review]

**Pull request merged**